File: 073_java_test/src/CalculatorTest.java
package com.example.calc;

/**
 * A test with a plain main() method. Bazel's test contract is just "exit 0 on
 * success", so no test framework is strictly required - which is handy when
 * you do not want to pull JUnit from Maven.
 */
public class CalculatorTest {
  private static int failures = 0;

  private static void check(String what, int got, int want) {
    if (got != want) {
      System.out.printf("FAIL %s: got %d, want %d%n", what, got, want);
      failures++;
    } else {
      System.out.printf("ok   %s%n", what);
    }
  }

  public static void main(String[] args) {
    check("add(2,3)", Calculator.add(2, 3), 5);
    check("factorial(0)", Calculator.factorial(0), 1);
    check("factorial(5)", Calculator.factorial(5), 120);

    try {
      Calculator.factorial(-1);
      System.out.println("FAIL factorial(-1): expected an exception");
      failures++;
    } catch (IllegalArgumentException expected) {
      System.out.println("ok   factorial(-1) throws");
    }

    if (failures > 0) {
      System.exit(1);
    }
    System.out.println("all java tests passed");
  }
}

File: 004_cc_test/calculator_test.cc
#include <cstdio>
#include <cstdlib>

#include "calculator.h"

static int failures = 0;

static void Expect(const char* what, int got, int want) {
  if (got != want) {
    printf("FAIL %s: got %d, want %d\n", what, got, want);
    failures++;
  } else {
    printf("ok   %s\n", what);
  }
}

int main() {
  Expect("Add(2,3)", Add(2, 3), 5);
  Expect("Add(-1,1)", Add(-1, 1), 0);
  Expect("Divide(10,2)", Divide(10, 2), 5);
  Expect("Divide(1,0)", Divide(1, 0), 0);
  // A test binary signals failure with a non-zero exit code.
  return failures == 0 ? 0 : 1;
}
